src/SpeedDial/SpeedDial.js

```diff
@@ -0,0 +1,221 @@
+      duration: theme.transitions.duration.shortest,
+    }),
+  },
+  fabHidden: {
+    transform: 'scale(0)',
+  },
+  fabIcon: {
+    transition: theme.transitions.create('transform', {
+      duration: theme.transitions.duration.shorter,
+      easing: theme.transitions.easing.easeInOut,
+    }),
+  },
+  fabIconOpen: {
+    transform: 'rotate(135deg)',
+  },
+  actions: {
+    display: 'flex',
+    flexDirection: 'column-reverse', // Display the first action at the bottom.
+    marginBottom: 16,
+  },
+  actionsClosed: {
+    transition: 'top 0s linear 0.2s',
+  },
+});
+
+class SpeedDial extends React.Component {
+  state = {
+    nextKey: 'up',
+    prevKey: 'down',
+  };
+
+  handleKeyDown = (event: SyntheticUIEvent<>) => {
+    const actions = findDOMNode(this.actions);
+    const fab = findDOMNode(this.fab);
+    const key = keycode(event);
+    const currentFocus = document.activeElement;
+
+    // If not actions, FAB must be focused.
+    if (currentFocus.parentElement.parentElement !== actions) {
+      if ((key === 'up' || key === 'down') && this.props.open) {
+        event.preventDefault();
+        actions.firstChild.firstChild.focus();
+
+        // This determines which key focuses the next / previous action.
+        // For example, if a visually impaired user presses down to select the first action
+        // (i.e. following DOM ordering), down will select the next action, and up the previous.
+        this.setState({ nextKey: key });
+        this.setState({ prevKey: key === 'up' ? 'down' : 'up' });
+      }
+    } else if (key === this.state.prevKey) {
+      event.preventDefault();
+      if (currentFocus.parentElement.previousElementSibling) {
+        currentFocus.parentElement.previousElementSibling.firstChild.focus();
+      } else {
+        findDOMNode(this.fab).focus();
+      }
+    } else if (key === this.state.nextKey) {
+      event.preventDefault();
+      if (currentFocus.parentElement.nextElementSibling) {
+        currentFocus.parentElement.nextElementSibling.firstChild.focus();
+      }
+    } else if (key === 'esc') {
+      fab.focus();
+      this.props.onClose(event, key);
+    }
+    if (this.props.onKeyDown) {
+      this.props.onKeyDown(event, key);
+    }
+  };
+
+  render() {
+    const {
+      children: childrenProp,
+      classes,
+      className: classNameProp,
+      hidden,
+      icon: iconProp,
+      label,
+      onKeyDown,
+      onClose,
+      open,
+      ...other
+    } = this.props;
+
+    const rootClassName = classNames(classes.root, classNameProp);
+    const buttonClassName = classNames(classes.fab, { [classes.fabHidden]: hidden });
+    const fabIconClassName = classNames(classes.fabIcon, { [classes.fabIconOpen]: open });
+    const actionsClassName = classNames(classes.actions, { [classes.actionsClosed]: !open });
+
+    // Filter the label for valid id characters.
+    const id = label.replace(/^[^a-z]+|[^\w:.-]+/gi, '');
```

Comment:
So... a unique label is required?

Reply:
Yes. The alternative was an explicit `id` prop, but I was trying to minimize the API surface. Given that you'd only ever have one SpeedDial visible, and if there's more than one they presumably have different purposes, it didn't seem like a major limitation, but I should probably add a note to the prop docs.